File: backend/app/schemas/paper_trading.py
"""Request/response contracts for the paper-trading API (Phase 3, M4).

Validated, bounded inputs (a bad symbol list, capital, or risk limit is a clean
4xx, never a 500) and JSON-safe response shapes for the deployment lifecycle, the
portfolio dashboard, and the live-vs-backtest comparison. Every read carries the
simulated-only disclaimer the product requires.
"""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

DISCLAIMER = (
    "Simulated paper trading against Alpaca's paper endpoint. No real money is "
    "traded. Results are not financial advice and do not imply real or future "
    "returns."
)

RunPhase = Literal["submit", "reconcile", "both"]


class DeploymentRiskConfig(BaseModel):
    """The portfolio sizing / cost / risk-limit knobs (everything except capital).

    Defaults follow the Phase 3 plan §3.3: no leverage (gross cap 1.0), ≤5 open
    positions, a −20% portfolio drawdown kill. Bounds reject leverage, negative
    costs, and nonsensical fractions before anything is persisted.
    """

    model_config = ConfigDict(extra="forbid")

    fee_bps: float = Field(default=5.0, ge=0.0, le=1000.0)
    slippage_bps: float = Field(default=5.0, ge=0.0, le=1000.0)
    target_vol: float | None = Field(default=None, gt=0.0, le=5.0)
    vol_lookback: int = Field(default=20, ge=2, le=512)
    max_position_pct: float = Field(default=0.95, gt=0.0, le=1.0)
    # Cash account, no leverage: gross exposure capped at 100% of equity.
    gross_exposure_cap: float = Field(default=1.0, gt=0.0, le=1.0)
    max_open_positions: int = Field(default=5, ge=1, le=100)
    per_order_notional_cap: float | None = Field(default=None, gt=0.0)
    stop_loss_pct: float | None = Field(default=None, gt=0.0, lt=1.0)
    take_profit_pct: float | None = Field(default=None, gt=0.0, le=10.0)
    max_drawdown_cutoff_pct: float | None = Field(default=0.20, gt=0.0, lt=1.0)


class DeploymentCreateRequest(BaseModel):
    """Create a paper-trading deployment."""

    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=128)
    strategy_name: str = Field(min_length=1)
    params: dict = Field(default_factory=dict)
    symbols: list[str] = Field(min_length=1, max_length=50)
    starting_capital: float = Field(gt=0.0)
    config: DeploymentRiskConfig = Field(default_factory=DeploymentRiskConfig)
    enabled: bool = True


class DeploymentUpdateRequest(BaseModel):
    """Patch a deployment's definition. Only provided fields change."""

    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=128)
    params: dict | None = None
    symbols: list[str] | None = Field(default=None, min_length=1, max_length=50)
    starting_capital: float | None = Field(default=None, gt=0.0)
    config: DeploymentRiskConfig | None = None


class EnableRequest(BaseModel):
    enabled: bool


class RunTriggerRequest(BaseModel):
    """Trigger a paper run. ``phase`` defaults to a full run-now."""

    trading_day: date | None = None
    phase: RunPhase = "both"


class RunTriggerResponse(BaseModel):
    job_id: str | None
    status: str
    deployment_id: int
    phase: RunPhase
    trading_day: date | None


class KillSwitchRequest(BaseModel):
    active: bool
    reason: str = ""


class KillSwitchStatus(BaseModel):
    active: bool
    reason: str


# --- Read models ------------------------------------------------------------


class DeploymentSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    strategy_name: str
    symbols: list[str]
    starting_capital: float
    enabled: bool
    status: str
    halt_reason: str | None
    created_at: datetime
    updated_at: datetime


class DeploymentDetail(DeploymentSummary):
    params: dict
    config: dict


class PortfolioSnapshotOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    trading_day: date
    equity: float
    cash: float
    position_value: float
    gross_exposure_pct: float
    drawdown_pct: float
    peak_equity: float
    num_positions: int


class PositionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    trading_day: date
    symbol: str
    quantity: float
    avg_entry_price: float
    market_value: float
    current_price: float


class OrderOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    trading_day: date
    symbol: str
    side: str
    intended_quantity: float
    intended_notional: float
    reference_price: float
    status: str
    filled_quantity: float
    reason: str
    submitted_at: datetime | None


class FillOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    trading_day: date
    symbol: str
    side: str
    quantity: float
    price: float
    modeled_reference_price: float
    slippage_delta: float
    filled_at: datetime | None


class ReconOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    trading_day: date
    kind: str
    symbol: str | None
    detail: str
    created_at: datetime


class SlippageSummary(BaseModel):
    """The implementation-shortfall distribution (realized fill minus modeled
    open, cost-signed) — the quantified residual backtest↔paper gap."""

    count: int
    mean: float
    median: float
    min: float
    max: float


class PortfolioView(BaseModel):
    """The paper-trading dashboard payload for one deployment."""

    deployment: DeploymentDetail
    equity_curve: list[PortfolioSnapshotOut]
    positions: list[PositionOut]
    orders: list[OrderOut]
    fills: list[FillOut]
    reconciliations: list[ReconOut]
    slippage: SlippageSummary
    global_kill: KillSwitchStatus
    disclaimer: str = DISCLAIMER


class MetricsOut(BaseModel):
    total_return_pct: float
    annualized_return_pct: float
    max_drawdown_pct: float
    sharpe_ratio: float
    sortino_ratio: float
    win_rate: float
    profit_factor: float
    num_round_trips: int


class ComparisonView(BaseModel):
    """The live paper results beside the deployment's backtested expectation on the
    same allocation logic, plus the measured residual fill-model gap."""

    deployment_id: int
    backtest_expectation: MetricsOut | None
    live_equity_curve: list[PortfolioSnapshotOut]
    slippage: SlippageSummary
    caveat: str = (
        "The backtest models a next-open fill; paper fills against real quotes "
        "and does not simulate dividends, market impact, latency, or queue "
        "position. The slippage distribution is the measured backtest↔paper gap; "
        "the paper↔live gap is larger and not modeled. Simulated only."
    )
    disclaimer: str = DISCLAIMER

File: frontend/src/types/paperTrading.ts
// Paper-trading types (Phase 3 M5), mirroring app/schemas/paper_trading.py.
// Simulated paper trading only — no real money is traded.

export interface DeploymentRiskConfig {
  fee_bps: number
  slippage_bps: number
  target_vol: number | null
  vol_lookback: number
  max_position_pct: number
  gross_exposure_cap: number
  max_open_positions: number
  per_order_notional_cap: number | null
  stop_loss_pct: number | null
  take_profit_pct: number | null
  max_drawdown_cutoff_pct: number | null
}

export interface DeploymentCreateRequest {
  name: string
  strategy_name: string
  params: Record<string, number>
  symbols: string[]
  starting_capital: number
  config: DeploymentRiskConfig
  enabled: boolean
}

export interface DeploymentSummary {
  id: number
  name: string
  strategy_name: string
  symbols: string[]
  starting_capital: number
  enabled: boolean
  status: string
  halt_reason: string | null
  created_at: string
  updated_at: string
}

export interface DeploymentDetail extends DeploymentSummary {
  params: Record<string, number>
  config: DeploymentRiskConfig
}

export interface PortfolioSnapshot {
  trading_day: string
  equity: number
  cash: number
  position_value: number
  gross_exposure_pct: number
  drawdown_pct: number
  peak_equity: number
  num_positions: number
}

export interface Position {
  trading_day: string
  symbol: string
  quantity: number
  avg_entry_price: number
  market_value: number
  current_price: number
}

export interface Order {
  id: number
  trading_day: string
  symbol: string
  side: string
  intended_quantity: number
  intended_notional: number
  reference_price: number
  status: string
  filled_quantity: number
  reason: string
  submitted_at: string | null
}

export interface Fill {
  id: number
  trading_day: string
  symbol: string
  side: string
  quantity: number
  price: number
  modeled_reference_price: number
  slippage_delta: number
  filled_at: string | null
}

export interface Recon {
  id: number
  trading_day: string
  kind: string
  symbol: string | null
  detail: string
  created_at: string
}

export interface SlippageSummary {
  count: number
  mean: number
  median: number
  min: number
  max: number
}

export interface KillSwitchStatus {
  active: boolean
  reason: string
}

export interface PortfolioView {
  deployment: DeploymentDetail
  equity_curve: PortfolioSnapshot[]
  positions: Position[]
  orders: Order[]
  fills: Fill[]
  reconciliations: Recon[]
  slippage: SlippageSummary
  global_kill: KillSwitchStatus
  disclaimer: string
}

export interface MetricsOut {
  total_return_pct: number
  annualized_return_pct: number
  max_drawdown_pct: number
  sharpe_ratio: number
  sortino_ratio: number
  win_rate: number
  profit_factor: number
  num_round_trips: number
}

export interface ComparisonView {
  deployment_id: number
  backtest_expectation: MetricsOut | null
  live_equity_curve: PortfolioSnapshot[]
  slippage: SlippageSummary
  caveat: string
  disclaimer: string
}

export interface RunTriggerResponse {
  job_id: string | null
  status: string
  deployment_id: number
  phase: string
  trading_day: string | null
}
